File: core/src/main/java/io/github/bindglam/neko/registry/DirectScalableRegistry.java
package io.github.bindglam.neko.registry;

import net.kyori.adventure.key.Key;
import org.jetbrains.annotations.NotNull;

public final class DirectScalableRegistry<T> extends ScalableRegistry<T> implements DirectWritableRegistry<T> {

    @Override
    public @NotNull T register(@NotNull Key key, @NotNull T value) {
        if (isLocked) {
            throw new IllegalStateException("The registry is locked");
        }
        if (map.containsKey(key.asString())) {
            throw new IllegalStateException("The registry already contains the key");
        }
        map.put(key.asString(), value);
        return value;
    }
}

File: core/src/main/kotlin/io/github/bindglam/neko/registry/DirectScalableRegistry.kt
package io.github.bindglam.neko.registry

import net.kyori.adventure.key.Key

class DirectScalableRegistry<T> : ScalableRegistry<T>(), DirectWritableRegistry<T> {
    override fun register(key: Key, value: T & Any): T & Any {
        if(isLocked)
            error("The registry is locked")
        if(map.contains(key.asString()))
            error("The registry already contains the key")
        return value.also { map[key.asString()] = it }
    }
}